{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import argparse\n",
    "import os\n",
    "import subprocess\n",
    "import io\n",
    "import argparse\n",
    "from sklearn.ensemble import IsolationForest\n",
    "import plotly\n",
    "import plotly.graph_objs as go\n",
    "\n",
    "\n",
    "###########################################\n",
    "#               FUNCTIONS                 #\n",
    "###########################################\n",
    "\n",
    "def generateReport(final, output_dir, reads):\n",
    "    if not os.path.isdir(output_dir):\n",
    "        cmd = [\"mkdir\", output_dir]\n",
    "        subprocess.check_output(cmd)\n",
    "\n",
    "    cmd = [\"cp\", \"-r\", \"/Users/admin/Documents/CNV/ressources\", output_dir]\n",
    "    subprocess.check_output(cmd)\n",
    "    \n",
    "    output_report = output_dir+\"/run.html\"\n",
    "    html = getTemplate()\n",
    "\n",
    "    for i in final.index:\n",
    "        fff=final.loc[i]\n",
    "\n",
    "        run = fff['Run']\n",
    "        sample = fff['Sample name']\n",
    "        region = fff['Region']\n",
    "\n",
    "        generateGraph(sample, output_dir, reads, region)\n",
    "\n",
    "        ratio = str(round(fff['Reads ratio'],2))\n",
    "        score = str(round(fff['Score'],2))\n",
    "\n",
    "        html = html.replace('<!--ADD RUN-->','<!--ADD RUN-->'+'\\n'+'\\t'+'\\t'+'\\t'+'<!--ADD RUN-->')\n",
    "\n",
    "        torep = \"<tr url=\"+\"\\\"\"+sample+'_'+region+\".html\\\" \"+\"\"\"style=\"cursor: pointer;\"><td>\"\"\"+run+\"</td><td>\"\"\"+sample+\"</td><td>\"\"\"+region+\"</td><td>\"\"\"+ratio+\"</td><td>\"\"\"+score+\"</td></tr>\"\n",
    "        html = html.replace('<!--ADD RUN-->',torep,1)\n",
    "        with open(output_report, \"w\") as FH_out:\n",
    "            FH_out.write(html)\n",
    "\n",
    "def generateGraph(sample,output_dir,reads,region):\n",
    "    dfneg = np.mean(reads[CNVneg],axis=1)\n",
    "    dfpos = reads[sample]\n",
    "\n",
    "    df = np.log2(np.array(dfpos)/np.array(dfneg))\n",
    "\n",
    "    genes = [i.split(\"_\")[0] for i in reads.index]\n",
    "\n",
    "    col=[]\n",
    "    for i in reads.index:\n",
    "        if i.startswith(region):\n",
    "            col.append(\"red\")\n",
    "        else:\n",
    "            col.append(\"black\")\n",
    "\n",
    "    data = [go.Scatter(\n",
    "        x = genes,\n",
    "        y = df,\n",
    "        mode = 'markers',\n",
    "        marker=dict(color=col,size=10),\n",
    "    )]\n",
    "    layout = go.Layout(\n",
    "            yaxis=dict(\n",
    "                title='Nomalized reads ratio',  \n",
    "            ),\n",
    "            plot_bgcolor='rgb(240,240,240)'\n",
    "        )\n",
    "    fig = go.Figure(data=data, layout=layout)\n",
    "    plotly.offline.plot(fig,filename=output_dir+'/'+sample+'_'+region+'.html', auto_open = False)\n",
    "\n",
    "\n",
    "def getTemplate():\n",
    "    html = \"\"\"\n",
    "<!DOCTYPE html>\n",
    "<html>\n",
    "  <head>\n",
    "    <title>ifCNV</title>\n",
    "    <meta charset=\"UTF-8\" />\n",
    "    <meta name=\"viewport\" content=\"width=device-width, initial-scale=1\" />\n",
    "    <link rel=\"stylesheet\" href=\"ressources/lib///w3.css\" />\n",
    "    <link rel=\"stylesheet\" href=\"ressources/fontawesome-free-5.3.1-web/css/all.css\" />\n",
    "    <link rel=\"stylesheet\" href=\"ressources/lib///datatables/datatables.min.css\" />\n",
    "    <link rel=\"stylesheet\" href=\"ressources/lib///interface.css\" />\n",
    "  </head>\n",
    "  <body>\n",
    "    <!-- Top container -->\n",
    "    <div class=\"w3-bar w3-top w3-blue-grey w3-large\" style=\"margin-left:10px;\">\n",
    "      <img src=\"ressources/ifCNV.png\" width=\"170\" height=\"80\">\n",
    "      <span style=\"color:white\" class=\"w3-display-middle\"><b><i><font size=\"+20\" face=\"calibri\">ifCNV Report</font></i></b></span>\n",
    "    </div>\n",
    "  <body>\n",
    "    <div id=\"Summary\" class=\"w3-container\" style=\"margin-left:50px;margin-top:125px;\">\n",
    "      <h2>\n",
    "        <b><center>Results</center></b>\n",
    "      </h2>\n",
    "\n",
    "    <div class=\"w3-main\" style=\"margin-left:100px;margin-right:100px;\">\n",
    "        <div id=\"runs\" class=\"w3-container\" style=\"padding-top:10px;\">\n",
    "          <table id=\"T_runs\" class=\"w3-table-all w3-hoverable w3-centered\">\n",
    "            <thead>\n",
    "              <tr>\n",
    "                <th>Run</th>\n",
    "                <th>Sample Name</th>\n",
    "                <th>Region</th>\n",
    "                <th>Reads ratio</th>\n",
    "                <th>Score</th>\n",
    "              </tr>\n",
    "            </thead>\n",
    "            <tbody>\n",
    "\t\t\t\t<!--ADD RUN-->\n",
    "            </tbody>\n",
    "          </table>\n",
    "        </div>\n",
    "    </div>\n",
    "</body>\n",
    "\n",
    "<script type=\"text/javascript\" src=\"ressources/lib/datatables/datatables.min.js\"></script>\n",
    "  <script>\n",
    "  $('body').on('mousedown', 'tr[url]', function(e){\n",
    "      var click = e.which;\n",
    "      var url = $(this).attr('url');\n",
    "      if(url){\n",
    "          if(click == 2 || (click == 1 && (e.ctrlKey || e.metaKey))){\n",
    "              window.open(url, '_blank');\n",
    "              window.focus();\n",
    "          }\n",
    "          else if(click == 1){\n",
    "              window.location.href = url;\n",
    "          }\n",
    "          return true;\n",
    "      }\n",
    "  });\n",
    "\n",
    "            $(document).ready(function() {\n",
    "                $('#T_runs tfoot th').each( function () {\n",
    "                    var title = $(this).text();\n",
    "                    $(this).html( '<input type=\"text\" placeholder=\"Search '+title+'\" style=\"width:100%\" />' );\n",
    "                } );\n",
    "\n",
    "                var table = $('#T_runs').DataTable({\n",
    "                    paging:true,\n",
    "                    order: [[1, 'desc']],\n",
    "                });\n",
    "\n",
    "                table.columns().every( function () {\n",
    "                    var that = this;\n",
    "\n",
    "                    $( 'input', this.footer() ).on( 'keyup change', function () {\n",
    "                        if ( that.search() !== this.value ) {\n",
    "                            that\n",
    "                                .search( this.value )\n",
    "                                .draw();\n",
    "                        }\n",
    "                    });\n",
    "                });\n",
    "            });\n",
    "        </script>\n",
    "\n",
    "</html>\n",
    "    \"\"\"\n",
    "    return(html)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def clean_reference(ref,outliers):\n",
    "    for i in outliers:\n",
    "        ref = ref.drop(labels=i,axis=1)\n",
    "\n",
    "    return ref\n",
    "\n",
    "def createReadsMatrix(pathToBam, bedFile, pathToBedtools, output=None, verbose=False):\n",
    "    cmd = [\"ls\", pathToBam]\n",
    "    res = subprocess.check_output(cmd)\n",
    "    final=pd.DataFrame()\n",
    "    \n",
    "    for i in res.decode('utf-8').split(\"\\n\"):\n",
    "        if i.endswith(\".bam\"):\n",
    "            if verbose==True:\n",
    "                print(\"Processing sample \"+i[:-4]+\"...\")\n",
    "            try:\n",
    "                command = [\n",
    "                pathToBedtools,\n",
    "                \"multicov\",\n",
    "                \"-bams\", pathToBam+\"/\"+i,\n",
    "                \"-bed\", bedFile]\n",
    "                res = subprocess.check_output(command)\n",
    "                data = io.StringIO(res.decode(\"utf-8\"))\n",
    "                df = pd.read_csv(data, sep='\\t',header=None)\n",
    "                nam = i[:-4]\n",
    "                final[nam] = df[len(df.columns)-1]\n",
    "                if verbose==True:\n",
    "                    print(i[:-4]+\" Done\")\n",
    "            except subprocess.CalledProcessError:\n",
    "                print(i[:-4] + \": skipped\")\n",
    "                print(\"Hint: Index file must be present in the folder\")\n",
    "\n",
    "\n",
    "\n",
    "    final.index = list(df[3])\n",
    "\n",
    "    if output is not None:\n",
    "        if verbose==True:\n",
    "            print(\"Reads matrix created !\")\n",
    "        final.to_csv(output,sep=\"\\t\")\n",
    "\n",
    "    return(final)\n",
    "\n",
    "\n",
    "def filterReads(reads,N,regtar=None,regsamp=None):\n",
    "    col = reads.columns\n",
    "    rows = reads.index\n",
    "    if regtar is not None:\n",
    "        reads = reads.filter(regex=regtar,axis=0)\n",
    "    if regsamp is not None:\n",
    "        reads = reads.filter(regex=\"^(?!\"+regsamp+\")\")\n",
    "    reads = reads.filter(regex=\"^(?!MSI)\",axis=0)\n",
    "    reads = reads.filter(regex=\"^(?!TN)\")\n",
    "    reads = reads.filter(regex=\"^(?!TP)\")\n",
    "    reads = reads.filter(regex=\"^(?!HD)\")\n",
    "    reads = reads.filter(regex=\"^(?!H2)\")\n",
    "    reads = reads.loc[reads.sum(axis=1)/len(reads.columns)>N,:]\n",
    "    filtered_samples = col[~np.in1d(col,reads.columns)]\n",
    "    filtered_targets = rows[~np.in1d(rows,reads.index)]\n",
    "    return(reads, filtered_samples, filtered_targets)\n",
    "\n",
    "\n",
    "def normalizeReads(reads):\n",
    "    reads_norm=reads/reads.sum(axis=0)\n",
    "    return(reads_norm)\n",
    "\n",
    "\n",
    "def aberrantSamples(reads,conta='auto'):    \n",
    "    tmp = np.percentile(reads, 99, axis = 0)/np.mean(reads, axis = 0)\n",
    "    random_data = np.array(tmp).reshape(-1,1)\n",
    "    clf = IsolationForest(contamination=conta).fit(random_data)\n",
    "    preds = clf.predict(random_data)\n",
    "    res_amp = np.array(reads.columns)[preds==-1]\n",
    "    \n",
    "    tmp = np.percentile(reads, 1, axis = 0)/np.mean(reads, axis = 0)\n",
    "    random_data = np.array(tmp).reshape(-1,1)\n",
    "    clf = IsolationForest(contamination=conta).fit(random_data)\n",
    "    preds = clf.predict(random_data)\n",
    "    res_del = np.array(reads.columns)[preds==-1]\n",
    "    \n",
    "    res = np.unique(np.concatenate((res_amp,res_del)))\n",
    "    norm = np.array(reads.columns[~np.in1d(reads.columns,res)])\n",
    "    \n",
    "    return(res, norm)\n",
    "\n",
    "\n",
    "\n",
    "def aberrantAmpliconsPerSample(name,reads_norm,CNVneg,conta=0.01):\n",
    "    random_data = np.array(reads_norm[name])#.reshape(-1,1)\n",
    "    norm = np.array(np.mean(reads_norm[CNVneg], axis = 1))\n",
    "    df = np.log2(random_data/norm)\n",
    "    clf = IsolationForest(contamination=conta).fit(df.reshape(-1,1))\n",
    "    preds = clf.predict(df.reshape(-1,1))\n",
    "    return(np.array(reads_norm.index)[preds==-1])\n",
    "\n",
    "\n",
    "def scoreAmplif(k,n,N):\n",
    "    p = n/N\n",
    "    x = np.log(1/((p**k)*(1-p)**(n-k)))*(k/n)\n",
    "    return x\n",
    "\n",
    "\n",
    "def amplifEvalGene(reads_norm,region,CNVneg,sample):\n",
    "    reads_m = reads_norm.filter(regex=\"^\"+region,axis=0)\n",
    "    pos = np.array(reads_m[sample])\n",
    "    norm = np.array(np.mean(reads_m[CNVneg], axis = 1))\n",
    "    df = pos/norm\n",
    "    val = np.mean(df)\n",
    "    if val==np.inf:\n",
    "        val = 100\n",
    "    return val\n",
    "\n",
    "\n",
    "def aberrantAmpliconsFinal(reads, reads_norm, CNVpos, CNVneg, scoreThreshold=10,conta=0.01,mode=\"fast\",run=\"ifCNV\"):\n",
    "    f = pd.DataFrame(columns=[\"Run\",\"Sample name\",\"Region\",\"Reads ratio\",\"Score\"])\n",
    "        \n",
    "    if mode==\"extensive\":\n",
    "        samples = [*CNVpos,*CNVneg]\n",
    "    if mode==\"fast\":\n",
    "        samples = CNVpos\n",
    "\n",
    "    q=0\n",
    "    for name in samples:\n",
    "        abAmp = aberrantAmpliconsPerSample(name,reads_norm,CNVneg,conta=conta)\n",
    "        if abAmp.shape!=(0,):\n",
    "            genes = np.unique([i.split('_')[0] for i in abAmp])\n",
    "            for gene in genes:\n",
    "                r = re.compile(gene)\n",
    "                abEx = list(filter(r.match, abAmp))\n",
    "                exons1 = [i.split('_')[0]+\"_\"+i.split('_')[1] for i in abEx]\n",
    "                tmp = reads.filter(regex=\"^\"+gene,axis=0)\n",
    "                exons2 = [i.split('_')[0]+\"_\"+i.split('_')[1] for i in tmp.index]\n",
    "\n",
    "                score = scoreAmplif(len(abEx),tmp.shape[0],reads.shape[0])\n",
    "                amplif = amplifEvalGene(reads_norm, gene, CNVneg, name)\n",
    "\n",
    "                if score>scoreThreshold:\n",
    "                    f.loc[q] = [run,name,gene,amplif,score]\n",
    "                    q=q+1\n",
    "\n",
    "    return(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing sample C21D00896...\n",
      "C21D00896: skipped\n",
      "Hint: Index file must be present in the folder\n",
      "Processing sample C21D00913...\n",
      "C21D00913 Done\n",
      "Processing sample C21D00920...\n",
      "C21D00920 Done\n",
      "Processing sample N21D00939...\n",
      "N21D00939 Done\n"
     ]
    }
   ],
   "source": [
    "pathBam = \"/Users/admin/Documents/tmp/bam_test/\"\n",
    "bed = \"/Users/admin/Documents/CNV/Panel_Juno_v3.bed\"\n",
    "bedtools = \"/Users/admin/miniconda3/bin/bedtools\"\n",
    "\n",
    "reads = createReadsMatrix(pathToBam=pathBam,bedFile=bed,pathToBedtools=bedtools,verbose=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "filteredReads, filteredS, filteredT = filterReads(reads, 200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "normReads = normalizeReads(filteredReads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "CNVpos, CNVneg = aberrantSamples(filteredReads,conta=\"auto\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Run Sample name  Region  Reads ratio     Score\n",
      "0  ifCNV   C21D00913    TERT     6.974571  2.888331\n",
      "1  ifCNV   N21D00939  CDKN2A     0.060880  3.072586\n",
      "2  ifCNV   N21D00939    EGFR    20.899706  4.593773\n",
      "3  ifCNV   N21D00939    PTEN     0.070813  3.117672\n",
      "4  ifCNV   N21D00939    RAC1     0.003601  7.311218\n"
     ]
    }
   ],
   "source": [
    "final = aberrantAmpliconsFinal(filteredReads,normReads,CNVpos,CNVneg,mode=\"fast\",scoreThreshold=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "generateReport(final,'/Users/admin/Documents/CNV/output_ifCNV/',normReads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>C21D00913</th>\n",
       "      <th>C21D00920</th>\n",
       "      <th>N21D00939</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>JAK1_E25_STA100319</th>\n",
       "      <td>1365</td>\n",
       "      <td>157</td>\n",
       "      <td>730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JAK1_E25_STA100320</th>\n",
       "      <td>1982</td>\n",
       "      <td>238</td>\n",
       "      <td>950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JAK1_E25_STA100321</th>\n",
       "      <td>1374</td>\n",
       "      <td>188</td>\n",
       "      <td>610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JAK1_E24_STA100322</th>\n",
       "      <td>5556</td>\n",
       "      <td>668</td>\n",
       "      <td>2735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JAK1_E24_STA100323</th>\n",
       "      <td>7208</td>\n",
       "      <td>822</td>\n",
       "      <td>3264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAPK1_E3_STA100153</th>\n",
       "      <td>4350</td>\n",
       "      <td>550</td>\n",
       "      <td>2600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAPK1_E3_STA100154</th>\n",
       "      <td>4350</td>\n",
       "      <td>550</td>\n",
       "      <td>2600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAPK1_E2_STA100155</th>\n",
       "      <td>8989</td>\n",
       "      <td>1374</td>\n",
       "      <td>4871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAPK1_E2_STA100156</th>\n",
       "      <td>17774</td>\n",
       "      <td>2523</td>\n",
       "      <td>10771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAPK1_E2_STA100157</th>\n",
       "      <td>15419</td>\n",
       "      <td>2176</td>\n",
       "      <td>9599</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1536 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    C21D00913  C21D00920  N21D00939\n",
       "JAK1_E25_STA100319       1365        157        730\n",
       "JAK1_E25_STA100320       1982        238        950\n",
       "JAK1_E25_STA100321       1374        188        610\n",
       "JAK1_E24_STA100322       5556        668       2735\n",
       "JAK1_E24_STA100323       7208        822       3264\n",
       "...                       ...        ...        ...\n",
       "MAPK1_E3_STA100153       4350        550       2600\n",
       "MAPK1_E3_STA100154       4350        550       2600\n",
       "MAPK1_E2_STA100155       8989       1374       4871\n",
       "MAPK1_E2_STA100156      17774       2523      10771\n",
       "MAPK1_E2_STA100157      15419       2176       9599\n",
       "\n",
       "[1536 rows x 3 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv(\"/Users/admin/Documents/CNV/readsMatrix_test.tsv\",sep=\"\\t\",index_col=0)\n",
    "data\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
